src/core/featurehistory.cpp

```diff
@@ -0,0 +1,448 @@
+{
+  if ( mProject->homePath().isNull() )
+    return;
+
+  mObservedLayerIds.clear();
+
+  addLayerListeners();
+}
+
+void FeatureHistory::addLayerListeners()
+{
+  const QList<QgsMapLayer *> layers = mProject->mapLayers().values();
+
+  for ( QgsMapLayer *layer : layers )
+  {
+    QgsVectorLayer *vl = qobject_cast<QgsVectorLayer *>( layer );
+
+    if ( !vl )
+    {
+      continue;
+    }
+
+    if ( mObservedLayerIds.contains( vl->id() ) )
+    {
+      continue;
+    }
+
+    // make sure we disconnect vector layer signal slots for these events if they are already present
+    disconnect( vl, &QgsVectorLayer::beforeCommitChanges, this, &FeatureHistory::onBeforeCommitChanges );
+    disconnect( vl, &QgsVectorLayer::afterCommitChanges, this, &FeatureHistory::onAfterCommitChanges );
+    disconnect( vl, &QgsVectorLayer::committedFeaturesAdded, this, &FeatureHistory::onCommittedFeaturesAdded );
+    disconnect( vl, &QgsVectorLayer::committedFeaturesRemoved, this, &FeatureHistory::onCommittedFeaturesRemoved );
+
+    connect( vl, &QgsVectorLayer::beforeCommitChanges, this, &FeatureHistory::onBeforeCommitChanges );
+    connect( vl, &QgsVectorLayer::afterCommitChanges, this, &FeatureHistory::onAfterCommitChanges );
+    connect( vl, &QgsVectorLayer::committedFeaturesAdded, this, &FeatureHistory::onCommittedFeaturesAdded );
+    connect( vl, &QgsVectorLayer::committedFeaturesRemoved, this, &FeatureHistory::onCommittedFeaturesRemoved );
+
+    mObservedLayerIds.insert( vl->id() );
+  }
+}
+
+
+void FeatureHistory::onBeforeCommitChanges()
+{
+  if ( mIsApplyingModifications )
+  {
+    return;
+  }
+
+  QgsVectorLayer *vl = qobject_cast<QgsVectorLayer *>( sender() );
+
+  if ( !vl )
+  {
+    return;
+  }
+
+  QgsVectorLayerEditBuffer *eb = vl->editBuffer();
+
+  if ( !eb )
+  {
+    return;
+  }
+
+  const QgsFeatureIds deletedFids = eb->deletedFeatureIds();
+  const QgsFeatureIds changedGeometriesFids = qgis::listToSet( eb->changedGeometries().keys() );
+  const QgsFeatureIds changedAttributesFids = qgis::listToSet( eb->changedAttributeValues().keys() );
+  // NOTE QgsFeatureIds underlying implementation is QSet, so no need to check if the QgsFeatureId already exists
+  QgsFeatureIds changedFids;
+
+  for ( const QgsFeatureId fid : deletedFids )
+    changedFids.insert( fid );
+
+  for ( const QgsFeatureId fid : changedGeometriesFids )
+    changedFids.insert( fid );
+
+  for ( const QgsFeatureId fid : changedAttributesFids )
+    changedFids.insert( fid );
+
+  // NOTE we read the features from the dataProvider directly as we want to access the old values.
+  // If we use the layer, we get the values from the edit buffer.
+  QgsFeatureIterator featuresIt = vl->dataProvider()->getFeatures( QgsFeatureRequest( changedFids ) );
+  QMap<QgsFeatureId, QgsFeature> modifiedFeatures;
+  QgsFeature f;
+
+  // ? is it possible to use the iterator in a less ugly way? something like normal `for ( QgsFeature &f : it ) {}`
```

Comment:
I'm used to this while form; beauty is in the eye of the beholder, I see no uglyness :wink:

Reply:
it's a matter of taste for sure.